Refactor scripts/format.py

adsharma opened the pull request:
The current format.py creates a temp file, formats it and attempts to rename.
However, `os.rename()` can fail if the src and target are on different filesystems.

An exception is thrown, but the paralllel formatting mechanism doesn't report it to the user.

Fixed by using in-place formatting.

Also uses argparse for argument parsing.

Test plan:

```
make format-check
make format-fix
```

For each of the file types, make small format changes. Then test with --check and --fix.

adsharma commented:
Files which are out of spec:

```
./scripts/format.py main --fix --all --noconfirm
git diff --stat | gh gist create
```

[Gist](https://gist.github.com/adsharma/07dbdfa91dcdb8b2eefebb3c6e5a52bd)

The build errors are because `cmake-format` is not capable of formatting these files:

```
-    ".test",
-    ".test_slow",
-    ".test_coverage",
-    ".benchmark",
```

I'll remove them from the list of extensions for now (unless there is evidence of cmake-format being able to format them)

Tishj commented on scripts/format.py:
```diff
@@ -1,467 +1,299 @@
-            if not is_old_header:
-                text += line
-
-    if ext == '.test' or ext == '.test_slow' or ext == '.test_coverage' or ext == '.benchmark':
```
If you're touching this file, you might want to preserve the existing behavior
With removing all of these lines I doubt that this behaves the same.

adsharma replied:
Does cmake-format actually have the ability to format `.test` files?

```
cmake-format -i test/planner/projection_binding.test
...
ValueError: Unexpected WORD token at 5:10, expecting l-paren, got 'ok'
```

adsharma replied:
But your overall observation is right - substantial parts of the file have been rewritten. Only the argparse/top level control flow can be described as a refactor.

adsharma replied:
I missed that. What if we move the formatting logic here to a tool of its own and invoke it similar to other tools?

adsharma replied:
Updated implementation tries to stick to the original algorithm as closely as possible. A small number of deltas were found and included in a separate commit for your review.

adsharma commented:
I noticed that old logic had a specific exception for:

```
        new_text = new_text.replace('ARGS &&...args', 'ARGS &&... args')
```

Did you take it up with `clang-format` folks to add a config option?

adsharma commented:
> the format script is not particularly critical to operation

That is exactly why I didn't bother to break up commits to that file. Hopefully we can verify the implementation via its externally documented behavior.

How about I break it up into:

* Split the test/benchmarking format code into it's own file. Executable as a script or importable as a function
* White space changes to a small number of files (4-5).
* Changes to format.py itself.

adsharma committed:
Whitespace only changes

These changes revert custom formatting changes in format.py, so
we could use `clang-format -i`.

The idea is to work with clang folks to work out a new config option
which gives us the desired behavior.

adsharma committed:
Move formatting logic for tests and benchmarks into a standalone script

adsharma committed:
Use inplace formatting

adsharma commented:
Please review the stack in this order:

#17349 
#17350 
#17305

adsharma committed:
Add missing newline which breaks clang-format

Tishj commented:
This is removing the changes made by <https://github.com/duckdb/duckdb/pull/6014>
I'd like to see reasoning for this, guarantees that the problem this fixed is no longer relevant.

Mytherin commented:
Agreed - we had this custom formatting rule in there because `clang-format` was inconsistent here (it would format in one way on some machines and in another on others) - leading to consistent issues. Undoing #6014 means all of those problems will likely appear again.

Can't we do a much more minimal change by just opening the target file and writing to it directly (i.e. doing the in-place write in the Python script)?

Mytherin commented:
My suggestion would be to split this into the following PRs:

* Switch argument handling to using `argparse`
* Switch to modifying lines inline by no longer writing to a temporary file but directly writing to the target file instead
* (Optionally) move the code for formatting tests to a separate script

adsharma commented:
> I'd like to see reasoning for this, guarantees that the problem this fixed is no longer relevant.

clang-format-11 was released Oct 2020. The current stable version is 20.x

Here's the [diff](https://gist.github.com/adsharma/d8747ff8d452862b246ac36cc5d1eda7) if were to upgrade clang-format. Doesn't look too disruptive.

adsharma commented:
> Can't we do a much more minimal change by just opening the target file and writing to it directly

I was trying to minimize the amount of error handling code to maintain instead of minimizing the size of the diff.
clang-format (or the other 3 tools) directly writing to the file means less error handling code in format.py (doesn't have to worry about partial writes etc).

```
git show 67ac603e7d --stat
 scripts/format.py | 685
 1 file changed, 285 insertions(+), 400 deletions(-)
 ```

adsharma commented:
Here's another reason why writing files from format.py is problematic. It doesn't preserve original file permissions:

```
diff --git a/scripts/format_test_benchmark.py b/scripts/format_test_benchmark.py
old mode 100755
new mode 100644
```

Seen via CI in #17350
